doc/user/operations.rst

```diff
@@ -0,0 +1,254 @@
+  this dictionary is the name of a slot of the compound operation. The value
+  associated with the key is a list of slots on the individual operations that
+  the compound slot corresponds to, in the form :samp:`{op-name}:{slot}`. In
+  this case we're indicating that the ``data`` slot on the fill operation and
+  the ``src`` slot on the reduction operation must point at the same buffer,
+  and from the outside that slot will be known as ``src``.
+
+  Sometimes the component operations might or might not have particular slots
+  depending on how they're configured. For convenience, non-existent slots are
+  silently ignored.
+
+  If a slot of a child is not listed here at all, it will be presented in the
+  parent as :samp:`{op-name}:{slot}`. It's recommended that this is avoided as
+  it exposes the internals of the implementation.
+
+The real power of :class:`.OperationSequence` is that it resolves all the
+padding requirements for the slots: any buffer bound to the ``src`` slot needs
+to satisfy the requirements of both the operations to which it is bound, but
+this is handled automatically.
+
+It is also possible for the child operations to themselves be instances of
+:class:`.OperationSequence` and thus create more and more complex operations,
+ending with a top-level operation that encapsulates an entire pipeline. At
+this level one may need to override the default :meth:`!__call__` or provide
+additional methods to run individual pieces of the pipeline as needed, while
+still having the benefit of linking slots together.
+
+Aliasing scratch buffers
+__________________________
+In more complex (particularly composed) operations, it's not uncommon for some
+slots to represent internal scratch space rather than inputs or outputs. When
+such operations are themselves composed, this can lead to wasted memory as
+each operation has its own scratch space, even though they are not in use at
+the same time and a single scratch area could be reused. If the scratch
+slots have the same shape and dtype then this can be handled by connecting
+them to a single slot on the parent operation, but that's not always the
+case.
+
+To handle this case, the :class:`.OperationSequence` constructor takes an
+additional dictionary for mapping slots, similar to the one already seen. The
+restrictions are much weaker: the child slots that are grouped together do not
+need to match in shape or dtype. The tradeoff is that the resulting slot is
+not fully-fledged (it is an :class:`.AliasIOSlot` rather than an
+:class:`.IOSlot`) so does not have a dtype, shape etc. It is normally not used
+directly, other than to serve as a source creating further alias slots if this
+operation is composed into an even higher-level one.
+
+It is recommended that you document scratch slots of your operations
+(including their sizes and anything unusual about data lifetime) so that users
+of your operation can decide how best to alias them with other scratch
+slots.
+
+Kernel fusion
```

Comment:
Extract from the CUDA 11 reveal article: 

"...some algorithms with a producer-consumer paradigm may observe performance benefits with persisting data in L2 between kernels, and therefore achieve higher bandwidth and performance. 

On A100, CUDA 11 offers API operations to set aside a portion of the 40-MB L2 cache to persist data accesses to global memory."

This seem like an alternative solution to kernel fusion. Not relevant for this review, just thought it was interesting.

Reply:
Yes, will be interesting to see if the much larger cache is actually useful. My experience has been that L2 hasn't been very useful for kernel-to-kernel reuse because kernels normally operate on large amounts of data (to get parallelism), except for heavily compute-bound kernels like matrix multiplies. Even 40MB is only 25 µs at 1.6 TB/s.